plugins/faustwp/includes/blocks/functions.php

```diff
@@ -1,70 +1,157 @@
+	if ( $error ) {
+		return $error;
 	}
 
-	// Check if it's a zip file.
-	if ( 'application/zip' !== $file['type'] ) {
-		return new \WP_Error( 'wrong_type', __( 'Not a zip file', 'faustwp' ) );
+	$dirs  = define_directories();
+	$error = ensure_directories_exist( $wp_filesystem, $dirs );
+	if ( $error ) {
+		return $error;
 	}
 
-	// Define directories.
-	$upload_dir = wp_upload_dir();
-	$target_dir = trailingslashit( $upload_dir['basedir'] ) . 'faustwp/';
-	$blocks_dir = $target_dir . 'blocks/';
-	$tmp_dir    = $target_dir . 'tmp_blocks/';
+	return process_and_replace_blocks( $wp_filesystem, $file, $dirs );
+}
 
-	// Ensure temporary directory exists.
-	if ( ! file_exists( $tmp_dir ) && ! wp_mkdir_p( $tmp_dir ) ) {
-		return new \WP_Error( 'mkdir_error', __( 'Could not create temporary directory', 'faustwp' ) );
+/**
+ * Processes the uploaded blockset file, unzips it, and replaces the old blocks directory.
+ *
+ * @param WP_Filesystem_Base $wp_filesystem Filesystem object.
+ * @param array              $file The uploaded file details.
+ * @param array              $dirs Directories array.
+ * @return WP_Error|bool True on success, WP_Error on failure.
+ */
+function process_and_replace_blocks( $wp_filesystem, $file, $dirs ) {
+	$target_file = $dirs['target'] . sanitize_file_name( basename( $file['name'] ) );
+
+	$move_result = move_uploaded_file( $wp_filesystem, $file, $target_file );
+	if ( is_wp_error( $move_result ) ) {
+		return $move_result;
 	}
 
-	// Move the uploaded file.
-	$target_file = $target_dir . sanitize_file_name( basename( $file['name'] ) );
-	if ( ! move_uploaded_file( $file['tmp_name'], $target_file ) ) {
-		return new \WP_Error( 'move_error', __( 'Could not move uploaded file', 'faustwp' ) );
+	$unzip_result = unzip_uploaded_file( $target_file, $dirs['blocks'] );
+	if ( is_wp_error( $unzip_result ) ) {
+		return $unzip_result;
 	}
 
-	// Unzip the file to the temporary directory.
-	if ( ! unzip_to_directory( $target_file, $tmp_dir ) ) {
-		rrmdir( $tmp_dir );  // Cleanup the temporary directory in case of unzip failure.
-		return new \WP_Error( 'unzip_error', __( 'Could not unzip the file', 'faustwp' ) );
+	cleanup_temp_directory( $wp_filesystem, $dirs['temp'] );
+
+	return true;
+}
+
+/**
+ * Validates the uploaded file type and readability.
+ *
+ * @param WP_Filesystem_Base $wp_filesystem Filesystem object.
+ * @param array              $file The uploaded file details.
+ * @return WP_Error|bool
+ */
+function validate_uploaded_file( $wp_filesystem, $file ) {
+	if ( 'application/zip' !== $file['type'] ) {
+		return new WP_Error( 'wrong_type', esc_html__( 'Not a zip file', 'faustwp' ) );
+	}
+
+	if ( ! $wp_filesystem->is_readable( $file['tmp_name'] ) ) {
+		return new WP_Error( 'file_read_error', esc_html__( 'Uploaded file is not readable', 'faustwp' ) );
 	}
 
-	// Replace the old blocks directory with the new content.
-	if ( is_dir( $blocks_dir ) ) {
-		rrmdir( $blocks_dir );
+	return true;
```

Comment:
If this always returns true then it will always fail in line 25 
https://github.com/wpengine/faustjs/pull/1656/files#r1396055606